tools/serverpod_cli/lib/src/analyzer/dart/definition_analyzers/class_analyzer.dart

```diff
@@ -0,0 +1,64 @@
+import 'package:analyzer/dart/element/element.dart';
+import 'package:path/path.dart' as path;
+import 'package:serverpod_cli/analyzer.dart';
+import 'package:serverpod_cli/src/analyzer/dart/definitions.dart';
+
+abstract class ClassAnalyzer {
+  static EndpointDefinition? analyze(
+    ClassElement element,
+    List<MethodDefinition> methodDefinitions,
+    CodeAnalysisCollector collector,
+    String filePath,
+    String rootPath,
+  ) {
+    var className = element.name;
+    var endpointName = _formatEndpointName(className);
+    var classDocumentationComment = element.documentationComment;
+    var subDirectoryParts = _getSubdirectoryParts(filePath, rootPath);
+    var endpointDef = EndpointDefinition(
+      name: endpointName,
+      documentationComment: classDocumentationComment,
+      className: className,
+      methods: methodDefinitions,
+      filePath: filePath,
+      subDirParts: subDirectoryParts,
+    );
+
+    return endpointDef;
+  }
+
+  static bool isEndpointClass(ClassElement element) {
+    if (element.supertype?.element.name != 'Endpoint') return false;
+
+    return true;
+  }
+
+  static String _formatEndpointName(String className) {
+    const removeEnding = 'Endpoint';
+
+    var endpointName = '${className[0].toLowerCase()}${className.substring(1)}';
+    if (endpointName.endsWith(removeEnding)) {
+      endpointName =
+          endpointName.substring(0, endpointName.length - removeEnding.length);
+    }
+
+    return endpointName;
+  }
+
+  static List<String> _getSubdirectoryParts(String filePath, String rootPath) {
+    // Get the subdirectory of the filePath by removing the first elements
+    // of the root path and the file path as long as they match.
+    var rootPathParts = path.split(rootPath);
+    var fileDirPathParts = path.split(path.dirname(filePath));
+    while (rootPathParts.isNotEmpty && fileDirPathParts.isNotEmpty) {
+      if (rootPathParts.first == fileDirPathParts.first) {
+        rootPathParts.removeAt(0);
+        fileDirPathParts.removeAt(0);
+      } else {
+        break;
+      }
+    }
+
+    return fileDirPathParts;
```

Comment:
Question: Isn't there some util method in the path package that does this exact thing already? I'm not 100% sure but I think I have seen similar code some where in our codebase.

Reply:
Yes, You can use `path.relative(...)` to do this. But it has some slight difference in behavior. It will give the relative path regardless if the path is a subfolder or not. And if is in the same folder it will give back `.` instead of empty list.

Do you think it would be better to use that instead?

Reply:
(The `.` return we could handle)

Reply:
It doesn't really matter, was more if there is something that solve the problem in the same way already it would be better to use it. You make the call on this one!

Reply:
In that case I'll leave it as is to prevent behavior changes in this PR. might revisit later.